src/pages/signup/SignupForm.tsx

```diff
@@ -0,0 +1,270 @@
+import { compose, graphql } from 'react-apollo';
+import * as Yup from 'yup';
+
+import { CreateUser, CREATE_OR_LOGIN_WITH_PROVIDERS, CREATE_USER, User } from 'networking/users';
+import { auth, login, signInWithGooglePopUp } from 'utils/firebase';
+import { getFriendlyErrorMessage } from 'utils/validators';
+
+interface SignupProps {
+  createOrLoginWithProviders: (id: string) => Promise<any>;
+  createUser: (user: CreateUser) => Promise<User>;
+}
+
+const LoginSchema = Yup.object().shape({
+  signupEmail: Yup.string()
+    .email('Please enter a valid email address.')
+    .required('Please enter an email address.'),
+  signupFirstName: Yup.string()
+    .required('Please enter your first name.'),
+  signupLastName: Yup.string()
+    .required('Please enter your last name.'),
+  signupPassword: Yup.string()
+    .min(8, 'Your password is too short, please enter a password that is at least 8 characters long')
+    .required('Please enter a valid password.'),
+  signupConfirmPassword: Yup.string()
+    .oneOf([Yup.ref('signupPassword')], 'Your passwords do not match.')
+    .required('Please re-enter your password here.'),
+});
+
+interface SignupFormInput {
+  signupEmail: string;
+  signupFirstName: string;
+  signupLastName: string;
+  signupPassword: string;
+  signupConfirmPassword: string;
+  authenticationError: string;
+}
+
+interface State {
+  providerError: string | null;
+}
+
+class SignupForm extends React.Component<SignupProps, State> {
+  readonly state = {
+    providerError: null,
+  };
+
+  render() {
+    return (
+      <Formik
+        initialValues={{
+          signupEmail: '',
+          signupFirstName: '',
+          signupLastName: '',
+          signupPassword: '',
+          signupConfirmPassword: '',
+          authenticationError: '',
+        }}
+        validationSchema={LoginSchema}
+        onSubmit={this.handleSubmit}
+      >
+        {({ errors, touched, setFieldValue, submitForm, isSubmitting }) => (
+          <Form method="POST">
+            <div className="mb-7">
+              <h2 className="h3 text-primary font-weight-normal mb-0">
+                Welcome to <span className="font-weight-semi-bold">Beenest</span>
+              </h2>
+              <p>Fill out the form to get started.</p>
+            </div>
+
+            <Row className="d-flex flex-sm-column flex-md-row">
+              <Col md={6}>
+                <FormGroup>
+                  <Label for="signupFirstName" className="form-label">
+                    First Name
+                  </Label>
+                  <Input
+                    type="text"
+                    name="signupFirstName"
+                    id="signupFirstName"
+                    tag={Field}
+                    onChange={(event: React.FormEvent<HTMLInputElement>) => {
+                      setFieldValue('signupFirstName', event.currentTarget.value);
+
+                      if (this.state.providerError !== null) {
```

Comment:
why is the conditional needed?

Reply:
added the conditional so it wouldnt set state if it was already null to prevent extra set states from leaking

Reply:
added the conditional because formik/reactstrap isnt good at handling server errors so we us our own state (vs formiks) to show an error from our server/firebase.  formik will reset the error fields on it's own but since formik has it's own state and validators it isnt hooked up to our server/firebase error state

Reply:
can this be a generic function so each of the inputs can re-use the function?
```
handleChange:
   setFieldValue(htmlId, event.currentTarget.value); 
...
```

Reply:
👍